ast/ksp/src/main/kotlin/me/tatarka/kotlin/ast/KSUtil.kt

```diff
@@ -26,10 +27,17 @@ internal fun KSAnnotated.hasAnnotation(packageName: String, simpleName: String):
 }
 
 private fun KSAnnotation.hasName(packageName: String, simpleName: String): Boolean {
-    // we can skip resolving if the short name doesn't match
-    if (shortName.asString() != simpleName) return false
     val declaration = annotationType.resolve().declaration
-    return declaration.packageName.asString() == packageName
+    return when {
+        shortName.asString() == simpleName -> declaration.packageName.asString() == packageName
+        declaration is KSTypeAlias -> declaration.isTypeAliasForName(packageName, simpleName)
+        else -> false
+    }
+}
+
+private fun KSTypeAlias.isTypeAliasForName(packageName: String, simpleName: String): Boolean {
+    val aliasedType = type.resolve()
+    return aliasedType.declaration.packageName.asString() == packageName && aliasedType.toString() == simpleName
 }
```

Comment:
You can have a typealis point to a typealias so this should probably recurse. I think you could also consolidate the 2 name checks by doing that.

```suggestion
    return when {
        declaration is KSTypeAlias -> declaration.type.resolve().declaration.hasName(packageName, simpleName)
        else ->  shortName.asString() == simpleName && declaration.packageName.asString() == packageName
    }
```

Reply:
Yes, you are right about the type alias pointing to type alias case. Fixed and added the test for it.  For now, I refactored the `isTypeAliasForName` function to start recursion with `tailerc` modifier to share the stack frame. I also changed the order of name checks in this function, to check the name of the aliased type first.

Reply:
And I do not know whether I get you right, but calling `hasName` like suggested here `declaration.type.resolve().declaration.hasName(packageName, simpleName)` on `declaration` after `resolve()` is not possible because we do not know if this `declaration` is a `KSAnnotation`, so I left the check for type alias in a separate function instead of trying to refactor for some generic solution, because it looked right and readable to me.

Reply:
Sorry yeah gave that suggestion without actually checking it was correct, this should work though:

```kotlin
private fun KSAnnotation.hasName(packageName: String, simpleName: String): Boolean {
    return annotationType.resolve().declaration.hasName(packageName, simpleName)
}

private tailrec fun KSDeclaration.hasName(packageName: String, simpleName: String): Boolean =
    when (this) {
        is KSTypeAlias -> type.resolve().declaration.hasName(packageName, simpleName)
        else -> this.simpleName.asString() == simpleName && this.packageName.asString() == packageName
    }
```

Reply:
@evant Now it works, changed as you suggested, thanks. Just to be sure, I see that there is also `Util.kt` file with similar implementation of `KSAnnotation.hasName` triggered from `KSAnnotated.hasAnnotation`, but as I check the implementation and usage, it is related to `@Provides` and `@Component` annotations, which I plan to tackle later on in separate PR, so I think we can let it be the way it is for now?

Reply:
Yeah that sounds right